<img src="./Img/Banner_1.png">

11 октября 2021 года<br>
ФРФТ20-2; ФРФТ20-1; ОД/207 (ул. Олеко Дундича, 23)

# 1. Основные понятия и определения

## 1.1. Основные понятия реляционных баз данных

Основой современных систем, использующих базы данных, является реляционная модель данных. В этой модели данные, представляющие информацию опредметно йобласти, организованы в видедвумерных таблиц, называемых отношениями.

<img src="./Img/img_01.png">

- Термины отношение и таблица обычно употребляются как синонимы. Верхняя строка таблицы-отношения называется заголовком
- Строки таблицы-отношения называются кортежами или записями. Столбцы называются атрибутами. Термины атрибут, столбец, колонка, поле обычно используются как синонимы. Каждый атрибут имеет имя, которое должно быть уникальным в конкретной таблице-отношении, однако в разных таблицах имена атрибутов могут совпадать
- Количество кортежей в таблице-отношении называется кардинальным числом отношения, а количество атрибутов называется степенью отношения
- Ключ или первичный ключ отношения – это уникальный идентификатор строк (кортежей), то есть такой атрибут (набор атрибутов), для которого в любой момент времени в отношении не существует строк с одинаковыми значениями этого атрибута (набора атрибутов). На приведенном рисунке таблицы ячейка с именем ключевого атрибута имеет нижнюю границу в видед войной черты
- Домен отношения – это совокупность значений, из которых могут выбираться значения конкретного атрибута

Отношения реляционной базы данных обладают следующими свойствами:
- в отношениях не должно быть кортежей-дубликатов;
- кортежи отношений неупорядочены;
- атрибуты отношений также неупорядочены.

Из уникальности кортежей следует, что в отношении всегда имеется атрибут или набор атрибутов, позволяющий идентифицировать кортеж; другими словами, в отношении всегда есть первичный ключ.

Из неупорядоченности кортежей следует, во-первых, что в отношении не существует другого способа адресации кортежей, кроме адресации по ключу; во-вторых, что в отношении не существует
таких понятий как первый кортеж, последний, предыдущий, следующий и т. п.

Из неупорядоченности атрибутов следует, что единственным способом их адресации в запросах является использование наименования атрибута.

<p style="color:#444444;"> <i><u>Замечание.</u></i> В отношениях, являющихся результатами запросов, SQL допускает наличие одинаковых строк. Для их устранения в запросе используется ключевое слово DISTINCT.</p>

Информация в реляционных базах данных, как правило, хранится не в одной таблице-отношении, а в нескольких. Для работы со связанными данными из нескольких таблиц важным является понятие так называемых внешних ключей. Внешним ключом таблицы называется атрибут или набор атрибутов этой таблицы, каждое значение которых в текущем состоянии таблицы всегда совпадает со значением атрибутов, являющихся ключом, в другой таблице. Внешние ключи используются для связывания значений атрибутов из разных таблиц. С помощью внешних ключей обеспечивается так называемая ссылочная целостность базы данных, т. е. согласованность данных, описывающих одни и те же объекты, но хранящихся в разных таблицах.

Язык SQL ориентирован на множества. Это означает, что в качестве входной информации для
формулируемого на языке SQL запроса к базе данных используется множество кортежей-записей одной или нескольких таблиц-отношений. В результате выполнения запроса также образуется множество кортежей результирующей таблицы-отношения.

## 1.2. Интерактивный и встроенный SQL
Существуют и используются две формы языка SQL: интерактивный SQL и встроенный SQL.  Интерактивный SQL используется для непосредственного ввода SQL-запросов пользователем и получения результата в интерактивном режиме.  Встроенный SQL состоит из команд SQL, встроенных внутрь программ, которые обычно написаны на некотором другом языке (Java, С, Python и др.)

## 1.3. Составные части SQL
В языке SQL выделяют язык определения и язык обработки данных.
- Язык определения данных – DDL (Data Definition Language), дает возможность создания, изменения и удаления различных объектов базы данных (таблиц, индексов, пользователей, привилегий и т. д.).
- Язык обработки данных – DML (Data Manipulation Language), предоставляет возможность выборки информации из базы данных и ее преобразования.

## 1.4. Типы данных SQL
В языке SQL имеются средства, позволяющие для каждого атрибута указывать тип данных, которому должны соответствовать все значения этого атрибута. Следует отметить, что определения типов данных в конкретных реализациях языка могут отличаться.

## 1.5. Используемые термины и обозначения

Ключевые слова – это используемые в выражениях SQL слова, имеющие специальное назначение (например, они могут обозначать конкретные команды SQL). Ключевые слова нельзя использовать для других целей, к примеру, в качестве имен объектов базы данных. Команды, или предложения, являются инструкциями, с помощью которых SQL обращается к базе данных.

## 1.6. Учебная база данных

<img src="./Img/img_02.png">
<img src="./Img/img_03.png">
<img src="./Img/img_04.png">
<img src="./Img/img_05.png">
<img src="./Img/img_06.png">
<img src="./Img/img_07.png">

Ответьте на вопросы
1. Какие поля приведенных таблиц являются первичными ключами? 
2. Какие данные хранятся в столбце 2 в таблице “Предмет обучения”?
3. Как по-другому называется строка? Столбец? 
4. Почему мы не можем запросить для просмотра первые пять строк?

# 2. Выборка данных (оператор SELECT)

## 2.1. Простейшие SELECT-запросы

Оператор SELECT (ВЫБРАТЬ) языка SQL является самым важным и самым часто используемым оператором.

`SELECT [DISTINCT] <список атрибутов>
    FROM <список таблиц> 
    [WHERE <условие выборки>]
    [ORDER BY <список атрибутов>]
    [GROUP BY <список атрибутов>]
    [HAVING <условие>]  
    [UNION <выражение с оператором SELECT>]`

Обязательным ключевым словом в предложении-запросе SELECT является слово FROM (из). За ключевым словом FROM указывается список разделенных запятыми имен таблиц.

Пример:
```
SELECT NAME, SURNAME 
    FROM STUDENT;
```
<img src="./Img/img_08.png">

Если необходимо вывести значения всех столбцов таблицы, то можно вместо перечисления их имен использовать символ “*” (звездочка)
```
SELECT *
    FROM STUDENT;
```

Получаемые в результате SQL-запроса таблицы не в полной мере отвечают определению реляционного отношения. В частности, в них могут оказаться кортежи с одинаковыми значениями атрибутов. Например, запрос “Получить список названий городов, где проживают студенты, сведения о которых находятся в таблице STUDENT”, можно записать в следующем виде

```
SELECT CITY FROM STUDENT;
```

Его результатом будет таблица

<img src="./Img/img_09.png">

Для исключения из результата SELECT-запроса повторяющихся записей используется ключевое слово DISTINCT (отличный). Запрос “Определить список названий различных городов, где проживают студенты, сведения о которых находятся в таблице STUDENT”, можно записать в следующем виде.

```
SELECT DISTINCT CITY
    FROM STUDENT;
```

Использование в операторе SELECT предложения, определяемого ключевым словом WHERE (где), позволяет задавать выражение условия (предикат), принимающее значение "Истина" или "Ложь" (а также неизвестно при использовании NULL) для значений полей строк таблиц, к которым обращается оператор SELECT. Предложение WHERE определяет, какие строки указанных таблиц должны быть выбраны. В таблицу, являющуюся результатом запроса, включаются только те строки, для которых условие принимает истинное значение.

Пример.
```
SELECT SURNAME,  NAME 
    FROM STUDENT
    WHERE SURNAME = "Петров";
```
<img src="./Img/img_10.png">

В задаваемых в предложении WHERE условиях могут использоваться операции сравнения, определяемые следующими операторами: = (равно), > (больше), < (меньше), >= (больше или равно), <= (меньше или равно), <> (не равно), а также логические операторы AND, OR и NOT. Например, запрос для получения имен и фамилий студентов, обучающихся на третьем курсе и получающих стипендию (размер стипендии больше нуля), будет выглядеть таким образом:
```
SELECT NAME, SURNAME
    FROM STUDENT
    WHERE (KURS = 3 AND STIPEND > 0);
```
<img src=./Img/icon-4017417__340.webp style="width:190px;height:190px">
<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания c помощью DB Browser for SQLite</p>

1. Напишите запрос к таблице SUBJECT, выводящий для каждой ее строки идентификатор (номер) предмета обучения, его наименование, семестр, в котором он читается, и количество отводимых на него часов.
2. Напишите запрос, позволяющий вывести все строки таблицы EXAM_MARKS, в которых предмет обучения имеет номер (SUBJ_ID), равный 12.  
3. Напишите запрос, выбирающий все данные из таблицы STUDENT, расположив столбцы таблицы в следующем порядке: KURS, SURNAME, NAME, STIPEND. 
4. Напишите запрос SELECT,  который выполняет вывод наименований предметов обучения (SUBJ_NAME) и следом за ним количества часов (HOUR) для каждого предмета обучения (SUBJECT) в 4-м семестре (SEMESTER).
5. Напишите запрос, позволяющий получить из таблицы EXAM_MARKS значения столбца MARK (экзаменационная оценка) для всех студентов, исключив из списка повторение одинаковы строк. 
6. Напишите запрос, который выполняет вывод списка фамилий студентов, обучающихся на третьем и более старших курсах. 
7. Напишите  запрос, выбирающий данные о фамилии, имении, номере курса для студентов, получающих стипендию больше 200.
8. Напишите запрос, выполняющий выборку из таблицы SUBJECT названий всех предметов обучения, на которые отводится более 30 часов.
9. Напишите запрос, который выполняет вывод списка университетов, рейтинг которых превышает 300 баллов.
10. Напишите запрос к таблице STUDENT для вывода списка фамилий (SURNAME), имен (NAME) и номера курса (KURS) всех студентов со стипендией большей или равной 100, и живущих в Воронеже.
11. Какие данные будут получены в результате выполнения запроса?
```
SELECT *
FROM STUDENT
WHERE (STIPEND < 100 OR
    NOT (BIRTHDAY >= "1990-03-10"
    AND STUDENT_ID > 59));
```
12. Какие данные будут получены в результате выполнения запроса?
```
SELECT *
    FROM STUDENT
    WHERE NOT ((BIRTHDAY = "1990-03-10" OR
        STIPEND > 100)
        AND STUDENT_ID >= 59)
```
13. Напишите запрос для получения списка студентов старше 25 лет, обучающихся на 1-м курсе.
14. Напишите запрос для получения списка предметов, для которых в 1-м семестре отведено более 100 часов.
15. Напишите запрос для получения списка преподавателей, живущих в Воронеже.
16. Напишите запрос для получения списка университетов, расположенных в Москве и имеющих рейтинг меньший, чем у ВГУ.
17. Напишите запрос для получения списка студентов, проживающих в Воронеже и не получающих стипендию.
18. Напишите запрос для получения списка студентов моложе 20 лет.
19. Напишите запрос для получения списка студентов без определенного места жительства.

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания 1-19 c помощью Python и библиотеки SQLAlchemy</p>

<p style="font-size:135%">Источники</p>

1. Астахова И.Ф., Мельников В.М., Толстобров А.П., Фертиков В.В. СУБД: язык SQL в примерах и задачах. — М.: ФИЗМАТЛИТ, 2009. — 168 с. — ISBN 978-5-9221-0816-4
2. SQLite — не игрушка. Электронный ресурс. https://habr.com/ru/post/547448/
3. SQL for Data Science. Электронный ресурс. https://medium.com/analytics-vidhya/sql-for-data-science-259298a10b0d
4. Официальный сайт SQLite. https://sqlite.org/

<img src="./Img/Banner_1.png">

14 октября 2021 года<br>
ФРФТ20-2; ФРФТ20-1; ОД/207 (ул. Олеко Дундича, 23)

## 2.2. Операторы IN, BETWEEN, LIKE, IS NULL

Операторы IN (равен любому из списка) и NOT IN (не равен ни одному из списка) используются для сравнения проверяемого значения поля с заданным списком. Этот список значений указывается в скобках справа от оператора IN. 

Пример.<br> Получить из таблицы EXAM_MARKS сведения о студентах, имеющих экзаменационные оценки только 4 и 5.
```
SELECT *
    FROM EXAM_MARKS
    WHERE MARK IN (4, 5);
```

Пример.<br> Получить сведения о студентах, не имеющих ни одной экзаменационной оценки 4 или 5.
```
SELECT *
    FROM EXAM_MARKS
    WHERE MARK NOT IN (4, 5)
```

Список значений не обязательно задается в явном виде, он может представлять собой результат подзапроса.
```
SELECT *
    FROM student
    WHERE city in (select city from lecturer);
```

Оператор BETWEEN используется для проверки условия вхождения значения поля в заданный интервал, т. е. вместо списка значений атрибута этот оператор задает границы его изменения.

Пример.<br>Получить сведения о предметах обучения, для которых количество отводимых часов лежит в пределах между 34 и 40.
```
SELECT *
    FROM SUBJECT
    WHERE HOUR BETWEEN 34 AND 40;
```

Оператор LIKE применим только к символьным полям типа CHAR или VARCHAR. Этот оператор осуществляет просмотр строковых значений полей с целью определения, входит ли заданная в операторе LIKE подстрока (образец поиска) в символьную строку, являющуюся значением проверяемого поля.

можно применять шаблон искомого образца строки, использующий следующие символы:
- символ подчеркивания “_”, указанный в шаблоне образца, определяет возможность наличия в указанном месте одного любого символа;
- символ “%” допускает присутствие в указанном месте проверяемой строки последовательности любых символов произвольной длины.

Пример.<br>Написать запрос, выбирающий из таблицы STUDENT сведения о студентах, у которых фамилии начинаются на букву “Р”.
```
SELECT *
FROM STUDENT
WHERE SURNAME LIKE "Р%";
```

В операторах сравнения =, <, >, <=, >=, <> и операторах **IN**, **BETWEEN** и **LIKE** при использовании **NULL** в качестве операнда будет возвращаться также **NULL**. В связи с этим, для проверки содержимого поля на наличие (отсутствие) в нем пустого значения **NULL** следует использовать специально предназначенные для этого операторы **IS NULL** (ЯВЛЯЕТСЯ ПУСТЫМ) и **IS NOT NULL** (НЕ ЯВЛЯЕТСЯ ПУСТЫМ), а не выражения = **NULL** или <> **NULL**.

<img src=./Img/icon-4017417__340.webp style="width:190px;height:190px">

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания c помощью DB Browser for SQLite</p>

1. Напишите запрос, выполняющий вывод находящихся в таблице EXAM_MARKS номеров предметов обучения, экзамены по которым сдавались в 2000 году.
2. Напишите запрос, выбирающий данные обо всех предметах обучения, экзамены по которым сданы студентами с идентификаторами 12 и 32.
3. Напишите запрос, который выполняет вывод названий предметов обучения, начинающихся на букву ‘И’.
4. Напишите запрос, выбирающий сведения о студентах, у которых имена начинаются на букву ‘И’ или ‘С’.
5. Напишите запрос для выбора из таблицы EXAM_MARKS записей, для которых отсутствуют значения оценок (поле MARK).
6. Напишите запрос, выполняющий вывод из таблицы EXAM_MARKS записей, для которых в поле MARK проставлены значения оценок для экзаменов в 2019 году.
7. Напишите запрос для получения списка преподавателей, проживающих в городах, в названиях которых присутствует дефис.
8. Напишите запрос для получения списка учебных заведений, в названиях которых использованы кавычки.
9. Напишите запрос для получения списка предметов, названия которых оканчиваются на ‘ия’.
10. Напишите запрос для получения списка учебных заведений, в названиях которых содержится слово ‘университет’.
11. Напишите запрос для получения списка студентов, фамилии которых начинаются на ‘Ков’ или на ‘Куз’.
12. Напишите запрос для получения списка предметов обучения, названия которых состоят из более одного слова.
13. Напишите запрос для получения списка учебных заведений, названия которых состоят как минимум из трех слов.
14. Напишите запрос для получения списка студентов, фамилии которых состоят из трех букв.

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания 1-14 c помощью Jupyter Notebook и языка програмирования Python</p>

In [1]:
#Ответы на задания раздела 2.2. Операторы IN, BETWEEN, LIKE, IS NULL
from sqlalchemy import MetaData, Table, create_engine, select, and_, or_, not_, cast, String
from sqlalchemy.sql import func
engine = create_engine('sqlite:///Data/Students.db')
metadata = MetaData()
metadata.reflect(bind=engine)
conn=engine.connect()
metadata.tables.keys()

exam_marks=Table("exam_marks", metadata, autoload=True, autoload_with="engine")
lecturer=Table("lecturer", metadata, autoload=True, autoload_with="engine")
student=Table("student", metadata, autoload=True, autoload_with="engine")
subj_lect=Table("subj_lect", metadata, autoload=True, autoload_with="engine")
subject=Table("subject", metadata, autoload=True, autoload_with="engine")
university=Table("university", metadata, autolad=True, autolad_with="engine")

In [36]:
#1 Способ 1, с помощью .between()
s=select([exam_marks.c.subj_id.distinct()])
s=s.where(exam_marks.c.exam_date.between("2000-01-01","2000-12-31"))
print(str(s)) # Если запрос сложный, то распечатываем выражение на языке SQL для самопроверки
print(s.compile().params)
result=conn.execute(s).fetchall()
print(result)

SELECT DISTINCT exam_marks.subj_id 
FROM exam_marks 
WHERE exam_marks.exam_date BETWEEN :exam_date_1 AND :exam_date_2
{'exam_date_1': '2000-01-01', 'exam_date_2': '2000-12-31'}
[(10,), (22,)]


In [37]:
#1 Способ 2, с помощью операторов >= , <=
s=select([exam_marks.c.subj_id.distinct()])
s=s.where(exam_marks.c.exam_date>="2000-01-01")
s=s.where(exam_marks.c.exam_date<="2000-12-31")
print(str(s)) # Если запрос сложный, то распечатываем выражение на языке SQL для самопроверки
print(s.compile().params)
result=conn.execute(s).fetchall()
print(result)

SELECT DISTINCT exam_marks.subj_id 
FROM exam_marks 
WHERE exam_marks.exam_date >= :exam_date_1 AND exam_marks.exam_date <= :exam_date_2
{'exam_date_1': '2000-01-01', 'exam_date_2': '2000-12-31'}
[(10,), (22,)]


In [39]:
#1 Способ 3, с помощью datetime()
from datetime import datetime
start_date = datetime(2000,1,1)
end_date = datetime(2000,12,31)
s=select([exam_marks.c.subj_id.distinct()])
s=s.where(and_(exam_marks.c.exam_date>=start_date, exam_marks.c.exam_date<=end_date))
result=conn.execute(s).fetchall()
print(result)

[(10,), (22,)]


In [47]:
#2
s=select([exam_marks.c.subj_id])
s=s.where(exam_marks.c.student_id.in_([12,32]))
result=conn.execute(s).fetchall()
print(result)

[(10,), (10,), (10,)]


In [50]:
#3
s=select([subject.c.subj_name])
s=s.where(subject.c.subj_name.startswith("И"))
result=conn.execute(s).fetchall()
print(result)

[('Информатика',), ('История',)]


In [53]:
#4
s=select([student.c.name, student.c.surname])
s=s.where(or_(student.c.name.startswith("И"),student.c.name.startswith("С")))
result=conn.execute(s).fetchall()
print(result)

[('Иван', 'Иванов'), ('Софья', 'Дементьева'), ('Ия', 'Соколова'), ('Ирина', 'Пак'), ('Синицин', 'Кондрат'), ('Светлана', 'Мишина'), ('Ирина', 'Сорокина'), ('Ирина', 'Жданова'), ('Ираида', 'Блохина'), ('Светлана', 'Мельникова'), ('Самуил', 'Никитин'), ('Икром', 'Назаров'), ('Софон', 'Захаров'), ('Иннокентий', 'Колобов'), ('Савва', 'Юдин'), ('Ипат', 'Комаров'), ('Семен', 'Симонов')]


In [80]:
#5 С помощью is_(None)
s=select([exam_marks])
s=s.where(exam_marks.c.mark.is_(None))
result=conn.execute(s).fetchall()
print(result)

[(639, 55, 10, None, datetime.datetime(2019, 5, 29, 0, 0))]


In [86]:
#6 С помощью isnot(None)
s=select([exam_marks])
s=s.where(exam_marks.c.exam_date.between("2019-01-01","2019-12-31"))
s=s.where(exam_marks.c.mark.isnot(None))
result=conn.execute(s).fetchall()
print(result, len(result))

[(10, 110, 22, 1, datetime.datetime(2019, 12, 25, 0, 0)), (12, 88, 56, 5, datetime.datetime(2019, 9, 2, 0, 0)), (14, 85, 13, 1, datetime.datetime(2019, 5, 29, 0, 0)), (16, 92, 10, 5, datetime.datetime(2019, 5, 29, 0, 0)), (18, 101, 94, 2, datetime.datetime(2019, 5, 26, 0, 0)), (23, 90, 22, 4, datetime.datetime(2019, 5, 29, 0, 0)), (25, 77, 12, 1, datetime.datetime(2019, 5, 29, 0, 0)), (28, 10, 12, 5, datetime.datetime(2019, 5, 29, 0, 0)), (30, 82, 56, 5, datetime.datetime(2019, 5, 29, 0, 0)), (31, 73, 11, 1, datetime.datetime(2019, 5, 29, 0, 0)), (95, 149, 13, 2, datetime.datetime(2019, 5, 29, 0, 0))] 11


In [87]:
#7
s=select([lecturer])
s=s.where(lecturer.c.city.like("%-%"))
result=conn.execute(s).fetchall()
print(result)

[(15, 'Прокофий', 'Дорофеев', 'Санкт-Петербург', 32)]


In [93]:
#8
s=select([university])
s=s.where(university.c.univ_name.contains("\""))
print(s, s.compile().params)
result=conn.execute(s).fetchall()
print(result)

SELECT university.univ_id, university.univ_name, university.rating, university.city 
FROM university 
WHERE (university.univ_name LIKE '%' || :univ_name_1 || '%') {'univ_name_1': '"'}
[(46, 'ФГБОУ "Политех"', 300, 'Санкт-Петербург')]


In [95]:
#9
s=select(subject.c.subj_name)
s=s.where(subject.c.subj_name.endswith("ия"))
result=conn.execute(s).fetchall()
print(result)

[('История',)]


In [97]:
#10
s=select([university])
s=s.where(university.c.univ_name.contains("У"))
result=conn.execute(s).fetchall()
print(result)

[(10, 'ВГУ', 296, 'Воронеж'), (11, 'НГУ', 345, 'Новосибирск'), (14, 'БГУ', 326, 'Белгород'), (15, 'ТГУ', 368, 'Томск'), (22, 'МГУ', 606, 'Москва'), (32, 'РГУ', 416, 'Ростов'), (44, 'ФинУ', 330, 'Москва'), (45, 'МГТУ', 372, 'Москва'), (46, 'ФГБОУ "Политех"', 300, 'Санкт-Петербург'), (47, 'КФУ', 330, 'Казань'), (48, 'УЛГУ', 231, 'Ульяновск')]


In [101]:
#11
s=select(student)
s=s.where(or_(student.c.surname.startswith("Ков"),student.c.surname.startswith("Куз")))
result=conn.execute(s).fetchall()
print(result)

[(10, 'Кузнецов', 'Борис', 0, 2, 'Брянск', datetime.datetime(2001, 12, 8, 0, 0), 10), (131, 'Ковалев', 'Ефим', 200, 2, 'Волгоград', datetime.datetime(1995, 11, 28, 0, 0), 47)]


In [213]:
#12
s=select(subject)
s=s.where(or_(subject.c.subj_name.contains(" "),subject.c.subj_name.contains("-")))
result=conn.execute(s).fetchall()
print(result)

[(11, 'Программирование на языках Python и SQL', 34, 2), (12, 'Анализ данных', 42, 1), (14, 'Обработка данных и моделирование в Excel', 36, 3), (23, 'Теория вероятностей и математическая статистика', 72, 4), (94, 'Английский язык', 56, 3)]


In [218]:
#13
s=select(university)
s=s.where(university.c.univ_name.like("% % %"))
result=conn.execute(s).fetchall()
print(result)

[(45, 'МГТУ им. Баумана', 372, 'Москва')]


In [220]:
#14
s=select(student.c.name,student.c.surname)
s=s.where(student.c.surname.like("___"))
result=conn.execute(s).fetchall()
print(result)

[('Елизавета', 'Дуб'), ('Ирина', 'Пак'), ('Оксана', 'Жук')]


## 2.3. Преобразование вывода и встроенные функции

В SQL реализованы операторы преобразования данных и встроенные функции, предназначенные для работы со значениями столбцов
и/или константами в выражениях. Использование этих операторов допустимо в запросах везде, где можно использовать выражения.

### Числовые, символьные и строковые константы

Если в запросе вместо спецификации столбца SQL обнаруживает число, то оно интерпретируется как *числовая константа*. Символьные константы должны указываться в одинарных кавычках. Если одинарная кавычка должна выводиться как часть строковой константы, то ее нужно предварить другой одинарной кавычкой.
```
SELECT 'Фамилия', SURNAME, 'Имя', NAME, 100
	FROM STUDENT;
```
### Арифметические операции для преобразования числовых данных

- Унарный (одиночный) оператор − (знак минус) изменяет знак числового значения, перед которым он стоит, на противоположный.
- Бинарные операторы +, −, * и / предоставляют возможность выполнения арифметических операций сложения, вычитания, умножения и деления.
```
SELECT SURNAME, NAME, STIPEND, KURS, -(STIPEND*KURS)/2
	FROM STUDENT
	WHERE KURS = 4 AND STIPEND > 0;
```

### Символьная операция конкатенации строк

Операция конкатенации, обозначаемая символом “||”, позволяет соединять (“склеивать”) значения двух или более столбцов символьного типа или символьных констант в одну строку.
```
SELECT surname || '_' || name, stipend
	FROM student
	WHERE kurs = 4 AND stipend > 200;
```

### Символьные строковые функции

**LTRIM** — удаление левых граничных символов.
```
LTRIM(<строка>[, <подстрока>])
```
- из <строки> удаляются слева символы, указанные в <подстроке>;
- если <подстрока> не указана, то по умолчанию удаляются пробелы;
- в <строку> справа добавляется столько пробелов, сколько символов слева было удалено, т. е. длина <строки> остается неизменной.

**RTRIM** — удаление правых граничных символов
```
RTRIM(<строка>[, <подстрока>]
```
Пример.
```
SELECT RTRIM(univ_name, "УГ") FROM university;
```

**SUBSTR** — выделение подстроки

```
SUBSTR(<строка>, <начало>[, <количество>]
```
- из <строки> выбирается заданное <количество> символов, начиная с указанной позиции в строке <начало>;
- если <количество> не задано, символы выбирают с <начала> и до конца <строки>.

Пример.
```
SELECT substr(subj_name,1,6) FROM subject;
```

**INSTR** — поиск подстроки
```
INSTR(<строка>, <подстрока>)
```
- тип возвращаемого значения — INT;
- функция возвращает позицию найденной подстроки.

**LENGTH** — определение длины строки
- длина <строки>, тип возвращаемого значения — INT;
- функция возвращает NULL, если <строка> имеет NULL-значение.

Пример.
```
SELECT subj_name, length(subj_name)
	FROM subject;
```
<img src=./Img/icon-4017417__340.webp style="width:190px;height:190px">
<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания c помощью DB Browser for SQLite</p>

1. Составьте запрос для таблицы STUDENT таким образом, чтобы в результат содержал только студентов, чья фамилия заканчивается на "ова" и чтобы выходная таблица содержала всего один столбец в следующем виде: В. Кудряшова; место жительства - Нижневартовск; день рождения 2002-02-18.
2. Составьте запрос, выводящий фамилии, имена студентов и величину получаемых ими стипендий, при этом значения стипендий должны быть увеличены в 100 раз.
3. Составьте запрос для таблицы UNIVERSITY таким образом, чтобы выходная таблица содержала всего один столбец в следующем виде: Код-10; ВГУ-г.ВОРОНЕЖ; Рейтинг=296.

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
Выполните задания 1-3 c помощью Jupyter Notebook и языка програмирования Python

In [30]:
#Ответы на задания раздела 2.3 Преобразование вывода и встроенные функции
from sqlalchemy import MetaData, Table, create_engine, select, and_, or_, not_, cast, String
from sqlalchemy.sql import func
engine = create_engine('sqlite:///Data/Students.db')
metadata = MetaData()
metadata.reflect(bind=engine)
conn=engine.connect()
metadata.tables.keys()

exam_marks=Table("exam_marks", metadata, autoload=True, autoload_with="engine")
lecturer=Table("lecturer", metadata, autoload=True, autoload_with="engine")
student=Table("student", metadata, autoload=True, autoload_with="engine")
subj_lect=Table("subj_lect", metadata, autoload=True, autoload_with="engine")
subject=Table("subject", metadata, autoload=True, autoload_with="engine")
university=Table("university", metadata, autolad=True, autolad_with="engine")

In [49]:
#1
s=select(func.substr(student.c.name,0,2)+"."+student.c.surname+"; "+
         "место жительства-"+student.c.city+"; "+"день рождения "+
         func.substr(cast(student.c.birthday,String()),0,11))
s=s.where(student.c.surname.like("%ова"))
result=conn.execute(s).fetchall()
print(result)

[('В.Кудряшова; место жительства-Нижневартовск; родилась-2002-02-18',), ('П.Рожкова; место жительства-Рубцовск; родилась-2001-09-15',), ('И.Соколова; место жительства-Колпашево; родилась-2002-06-25',), ('В.Семенова; место жительства-Вологда; родилась-2003-10-04',), ('В.Некрасова; место жительства-Иркутск; родилась-2004-08-23',), ('О.Казакова; место жительства-Кемерово; родилась-2000-03-29',), ('А.Миронова; место жительства-Владикавказ; родилась-2000-10-08',), ('Л.Одинцова; место жительства-Верхний Баскунчак; родилась-2002-03-07',), ('З.Игнатова; место жительства-Ростов; родилась-2000-09-26',), ('Д.Мельникова; место жительства-Архангельск; родилась-2002-08-01',), ('В.Фролова; место жительства-Кисловодск; родилась-1999-06-02',), ('Г.Шашкова; место жительства-Кущевская; родилась-2001-08-15',), ('М.Казакова; место жительства-Нальчик; родилась-1994-02-18',), ('Ф.Жукова; место жительства-Магадан; родилась-1998-10-06',), ('И.Жданова; место жительства-Тверь; родилась-2002-03-02',), ('С.Мельник

In [71]:
#2
s=select([student.c.surname, student.c.name,student.c.stipend*100])
result=conn.execute(s).fetchone()
print(result)

('Иванов', 'Иван', 15000)


In [72]:
#3
s=select("Код-"+university.c.univ_id+"; "+university.c.univ_name+"; "+"рейтинг="+university.c.rating)
result=conn.execute(s).fetchall()
print(result)

[(306,), (356,), (340,), (383,), (345,), (628,), (448,), (374,), (417,), (346,), (377,), (279,)]


In [74]:
s=select("Код-"+cast(university.c.univ_id,String())+"; "+
         university.c.univ_name+"; "+"рейтинг="+cast(university.c.rating,String()))
result=conn.execute(s).fetchall()
print(result)

[('Код-10; ВГУ; рейтинг=296',), ('Код-11; НГУ; рейтинг=345',), ('Код-14; БГУ; рейтинг=326',), ('Код-15; ТГУ; рейтинг=368',), ('Код-18; ВГМА; рейтинг=327',), ('Код-22; МГУ; рейтинг=606',), ('Код-32; РГУ; рейтинг=416',), ('Код-44; ФинУ; рейтинг=330',), ('Код-45; МГТУ; рейтинг=372',), ('Код-46; Политех; рейтинг=300',), ('Код-47; КФУ; рейтинг=330',), ('Код-48; УЛГУ; рейтинг=231',)]


<p style="font-size:135%">Источники</p>

1. Астахова И.Ф., Мельников В.М., Толстобров А.П., Фертиков В.В. СУБД: язык SQL в примерах и задачах. — М.: ФИЗМАТЛИТ, 2009. — 168 с. — ISBN 978-5-9221-0816-4
2. SQLite — не игрушка. Электронный ресурс. https://habr.com/ru/post/547448/
3. SQL for Data Science. Электронный ресурс. https://medium.com/analytics-vidhya/sql-for-data-science-259298a10b0d
4. Официальный сайт SQLite. https://sqlite.org/

<img src="./Img/Banner_1.png">

18 октября 2021 года<br>
ФРФТ20-2; ФРФТ20-1; ОД/207 (ул. Олеко Дундича, 23)


## 2.4. Агрегирование и групповые функции

Агрегирующие функции позволяют получать из таблицы сводную (агрегированную) информацию, выполняя операции над группой строк
таблицы. Для задания в SELECT-запросе агрегирующих операций используются следующие ключевые слова:
- COUNT определяет количество строк или значений поля, выбранных посредством запроса и не являющихся NULL-значениями;
- SUM вычисляет арифметическую сумму всех выбранных значений данного поля;
- AVG вычисляет среднее значение для всех выбранных значений данного поля;
- MAX вычисляет наибольшее из всех выбранных значений данного поля;
- MIN вычисляет наименьшее из всех выбранных значений данного поля.

```
SELECT min(rating), max(rating), round(avg(rating),2), count(univ_id)
	FROM university;
    
SELECT count(DISTINCT subj_id)
    FROM subject;
```

Предложение GROUP BY позволяет применять агрегирующие функции к каждой группе, определяемой общим значением поля или полей, указанных в этом предложении.
```
SELECT exam_marks.student_id, max(exam_marks.mark)
	FROM exam_marks
	GROUP by exam_marks.student_id;
```

В конструкции GROUP BY для группирования может быть использовано более одного столбца.
```
SELECT student_id, subj_id, max(mark)
    FROM exam_marks
    GROUP BY student_id, subj_id
```
Следует иметь в виду, что после ключевого слова SELECT должны быть использованы только те имена столбцов, которые указаны в
предложении GROUP BY.

Предложение HAVING определяет критерий, по которому группы следует включать в выходные данные (по аналогии с предложением
WHERE, которое осуществляет это для отдельных строк):

```
SELECT subj_name, max(HOUR)
	FROM subject
	GROUP BY subj_name
	HAVING max(hour) >= 50 AND max(hour)<70
```
<img src=./Img/icon-4017417__340.webp style="width:190px;height:190px">
<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания c помощью DB Browser for SQLite</p>

1. Напишите запрос для подсчета количества студентов, сдававших экзамен по предмету обучения с идентификатором 73.
2. Напишите запрос, который позволяет подсчитать в таблице EXAM_MARKS количество различных предметов обучения.
3. Напишите запрос, который для каждого студента выполняет выборку его идентификатора и минимальной из полученных им оценок.
4. Напишите запрос, который для каждого студента выполняет выборку его идентификатора и максимальной из полученных им оценок.
5. Напишите запрос, выполняющий вывод первой по алфавиту фамилии студента, начинающейся на букву "И".
6. Напишите запрос, который для каждого предмета обучения выводит наименование предмета и максимальное значение номера семестра, в котором этот предмет преподается.
7. Напишите запрос, который для каждого конкретного дня сдачи экзамена выводит данные о количестве студентов, сдававших экзамен в этот день.
8. Напишите запрос, выдающий средний балл для каждого студента.
9. Напишите запрос, выдающий средний балл для экзамена по каждому предмету.
10. Напишите запрос, определяющий количество сдававших студентов для каждого предмета (оценка больше двух).
11. Напишите запрос для определения количества предметов, изучаемых на каждом курсе.
12. Для каждого университета напишите запрос, выводящий суммарную стипендию обучающихся в нем студентов, с последующей сортировкой списка по этому значению.
13. Для каждого семестра напишите запрос, выводящий общее количество часов, отводимое на изучение соответствующих предметов.
14. Для каждого студента напишите запрос, выводящий среднее значение оценок, полученных им на всех экзаменах.
15. Для каждого студента напишите запрос, выводящий среднее значение оценок, полученных им по каждому предмету.
16. Напишите запрос, выводящий количество студентов, проживающих в каждом городе. Список отсортировать в порядке убывания количества студентов.
17. Для каждого университета напишите запрос, выводящий количество обучающихся в нем студентов, с последующей сортировкой списка по этому количеству.
18. Для каждого университета напишите запрос, выводящий количество работающих в нем преподавателей, с последующей сортировкой списка по этому количеству.
19. Для каждого университета напишите запрос, выводящий сумму стипендии, выплачиваемой студентам каждого курса.
20. Для каждого города напишите запрос, выводящий максимальный рейтинг университетов, в нем расположенных, с последующей сортировкой списка по значениям рейтингов.
21. Для каждого дня сдачи экзаменов напишите запрос, выводящий среднее значение всех экзаменационных оценок.
22. Для каждого дня сдачи экзаменов напишите запрос, выводящий максимальные оценки, полученные по каждому предмету.
23. Для каждого дня сдачи экзаменов напишите запрос, выводящий общее количество студентов, сдававших экзамены.
24. Для каждого дня сдачи экзаменов напишите запрос, выводящий общее количество экзаменов, сдававшихся каждым студентом.
25. Для каждого преподавателя напишите запрос, выводящий количество преподаваемых им предметов.
26. Для каждого предмета напишите запрос, выводящий количество преподавателей, ведущих по нему занятия.
27. Напишите запрос, выполняющий вывод списка студентов, имеющих только отличные оценки. Результат отстортируйте по идентификатору студента.
28. Напишите запрос, выполняющий вывод количества экзаменов, сданных (с положительной оценкой) студентами с идентификаторами 12, 32, 38.

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните указанные задания c помощью Python в интерактивной среде Jupyter notebook</p>

In [ ]:
# Ответы на задания раздела 2.4. Агрегирование и групповые функции

from sqlalchemy import MetaData, Table, create_engine, select, and_, or_, not_
from sqlalchemy.sql import func
engine = create_engine('sqlite:///Data/Students.db')
metadata = MetaData()
metadata.reflect(bind=engine)
conn=engine.connect()
metadata.tables.keys()

exam_marks=Table("exam_marks", metadata, autoload=True, autoload_with="engine")
lecturer=Table("lecturer", metadata, autoload=True, autoload_with="engine")
student=Table("student", metadata, autoload=True, autoload_with="engine")
subj_lect=Table("subj_lect", metadata, autoload=True, autoload_with="engine")
subject=Table("subject", metadata, autoload=True, autoload_with="engine")
university=Table("university", metadata, autolad=True, autolad_with="engine")

In [ ]:
#1
s=select([func.count(exam_marks.c.student_id)]).where(exam_marks.c.subj_id==73)
result=conn.execute(s)
result.scalar()

In [ ]:
#2
result=conn.execute(select([exam_marks.c.subj_id.distinct()])).fetchall()
print(len(result))

In [ ]:
#3
s=select([exam_marks.c.student_id,func.min(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.student_id)
result=conn.execute(s).fetchone() # Выводится одна запись для проверки, правильно fetchall()
print(result)

In [ ]:
#4
s=select([exam_marks.c.student_id,func.max(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.student_id)
result=conn.execute(s).fetchone() # Выводится одна запись для проверки, правильно fetchall()
print(result)

In [ ]:
#5
s=select([func.min(student.c.surname)]).where(student.c.surname.startswith("И"))
result=conn.execute(s).fetchone()
print(result)

In [ ]:
#6
s=select([subject.c.subj_name,func.max(subject.c.semester)])
s=s.group_by(subject.c.subj_name)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#7
s=select([exam_marks.c.exam_date, func.count(exam_marks.c.student_id)])
s=s.group_by(exam_marks.c.exam_date)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#8
s=select([exam_marks.c.student_id,func.avg(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.student_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#9
s=select([exam_marks.c.subj_id,func.avg(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.subj_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#10
s=select([exam_marks.c.subj_id,func.count(exam_marks.c.student_id)])
s=s.where(exam_marks.c.mark>2)
s=s.group_by(exam_marks.c.subj_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#11
s=select([(subject.c.semester+1)/2,func.count(subject.c.subj_name)])
s=s.group_by((subject.c.semester+1)/2)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#12
s=select([student.c.univ_id,func.sum(student.c.stipend)])
s=s.group_by(student.c.univ_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#13
s=select([subject.c.semester,func.sum(subject.c.hour)]).group_by(subject.c.semester)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#14
s=select([exam_marks.c.student_id,func.avg(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.student_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#15
s=select([exam_marks.c.student_id,exam_marks.c.subj_id,func.avg(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.student_id,exam_marks.c.subj_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#16
s=select([student.c.city,func.count(student.c.student_id)]).group_by(student.c.city)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#17
s=select([student.c.univ_id,func.count(student.c.univ_id)])
s=s.group_by(student.c.univ_id)
s=s.order_by(func.count(student.c.univ_id))
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#18
s=select([lecturer.c.univ_id, func.count(lecturer.c.lecturer_id)]).group_by(lecturer.c.univ_id)
s=s.order_by(func.count(lecturer.c.lecturer_id))
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#19
s=select([student.c.univ_id,student.c.kurs,func.sum(student.c.stipend)]).group_by(student.c.univ_id,student.c.kurs)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#20
s=select([university.c.city,func.max(university.c.rating)])
s=s.group_by(university.c.city)
s=s.order_by(func.max(university.c.rating))
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#21
s=select([exam_marks.c.exam_date,func.avg(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.exam_date)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#22
s=select([exam_marks.c.exam_date,exam_marks.c.subj_id,func.avg(exam_marks.c.mark)])
s=s.group_by(exam_marks.c.exam_date,exam_marks.c.subj_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#23
s=select([exam_marks.c.exam_date,func.count(exam_marks.c.student_id.distinct())])
s=s.group_by(exam_marks.c.exam_date)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#24
s=select([exam_marks.c.exam_date,exam_marks.c.student_id,func.count(exam_marks.c.exam_id)])
s=s.group_by(exam_marks.c.exam_date,exam_marks.c.student_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#25
s=select([subj_lect.c.lecturer_id,func.count(subj_lect.c.subj_id)])
s=s.group_by(subj_lect.c.lecturer_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#26
s=select([subj_lect.c.subj_id,subj_lect.c.lecturer_id])
s=s.group_by(subj_lect.c.subj_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#27
s=select([exam_marks.c.student_id.distinct()])
s=s.where(exam_marks.c.mark==5)
s=s.order_by(exam_marks.c.student_id)
result=conn.execute(s).fetchall()
print(result)

In [ ]:
#28 Способ 1: с использованием and_() и or_()
s=select([func.count(exam_marks.c.exam_id)])
s=s.where(and_(exam_marks.c.mark>2,or_(
    exam_marks.c.student_id==12,exam_marks.c.student_id==32,exam_marks.c.student_id==128
)))
result=conn.execute(s).fetchall()
print(result)

#28 Способ 2: с использованием and_() и in_()
s=select([func.count(exam_marks.c.exam_id)])
s=s.where(and_(exam_marks.c.mark>2,exam_marks.c.student_id.in_([12,32,128])))
result=conn.execute(s).fetchone()
print(result)

#28 Способ 3: с использованием in_()
s=select([func.count(exam_marks.c.exam_id)])
s=s.where(exam_marks.c.mark>2)
s=s.where(exam_marks.c.student_id.in_([12,32,128]))
result=conn.execute(s).scalar()
print(result)

<p style="font-size:135%">Источники</p>

1. Астахова И.Ф., Мельников В.М., Толстобров А.П., Фертиков В.В. СУБД: язык SQL в примерах и задачах. — М.: ФИЗМАТЛИТ, 2009. — 168 с. — ISBN 978-5-9221-0816-4
2. SQLite — не игрушка. Электронный ресурс. https://habr.com/ru/post/547448/
3. SQL for Data Science. Электронный ресурс. https://medium.com/analytics-vidhya/sql-for-data-science-259298a10b0d
4. Официальный сайт SQLite. https://sqlite.org/

<img src="./Img/Banner_1.png">

25 октября 2021 года<br>
ФРФТ20-2; ФРФТ20-1; ОД/207 (ул. Олеко Дундича, 23)

## 2.5. Неопределенные значения (NULL) в агрегирующих функциях

Отсутствие значений в полях таблицы, обозначенное ключевым
словом NULL, приводит к особенностям при выполнении агрегирующих
операций над данными, которые следует учитывать при их использо-
вании в SQL-запросах.

**Влияние NULL-значений в функции COUNT.** Если аргумент функции COUNT является константой или столбцом без пустых значений, то функция возвращает количество строк, к которым применимо определенное условие или группирование.

Если аргументом функции является столбец, содержащий пустое значение, то COUNT вернет число строк, которые не содержат пустые значения и к которым применимо определенное условие или группирование.

Поведение функции `COUNT(*)` не зависит от пустых значений. Она
возвратит общее количество строк в таблице.

**Влияние NULL-значений в функции AVG.** Среднее значение множества чисел равно сумме чисел, деленной на число элементов множества. Однако если некоторые элементы пусты, т. е. их значения неизвестны или не существуют, то деление на количество всех элементов множества приведет к неправильному результату.

Функция AVG вычисляет среднее значение всех известных значений множества элементов, т. е. эта функция подсчитывает сумму известных значений и делит ее на количество этих значений, а не на общее количество значений, среди которых могут быть NULL-значения. Если столбец состоит только из пустых значений, то функция AVG также возвратит NULL.


## 2.6. Результат действия трехзначных операторов

Условные операторы при отсутствии пустых значений возвращают либо TRUE (ИСТИНА), либо FALSE (ЛОЖЬ). Если же в столбце присутствуют пустые значения, то может быть возвращено и третье значение: UNKNOWN (НЕИЗВЕСТНО).

В этой схеме, например, условию `WHERE A=2`, где А — имя столбца, значения которого могут быть
неизвестны, при A=2 будет соответствовать TRUE, при A=4 в результате будет получено значение FALSE, а при отсутствующем значении A (NULL-значение) результат будет UNKNOWN. Пустые значения могут оказывать влияние на использование логических операторов NOT, AND и OR.

В следующей таблице показаны результаты применения оператора AND к двум логическим выражениям, когда одно выражение возвращает UNKNOWN.
<Table>
    <tr><th>Выражение 1</th><th>Выражение 2</th><th>Результат</th>
    <tr><td>TRUE</td><td>UNKNOWN</td><td>UNKNOWN</td>
    <tr><td>UNKNOWN</td><td>UNKNOWN</td><td>UNKNOWN</td>
    <tr><td>FALSE</td><td>UNKNOWN</td><td>FALSE</td>
</Table>
    
В следующей таблице показаны результаты применения оператора OR к двум логическим выражениям, когда одно выражение возвращает UNKNOWN.
<Table>
    <tr><th>Выражение 1</th><th>Выражение 2</th><th>Результат</th>
    <tr><td>TRUE</td><td>UNKNOWN</td><td>TRUE</td>
    <tr><td>UNKNOWN</td><td>UNKNOWN</td><td>UNKNOWN</td>
    <tr><td>FALSE</td><td>UNKNOWN</td><td>UNKNOWN</td>
</Table>

Для проверки наличия значения NULL в предложении WHERE запроса используются ключевые слова IS NULL или IS NOT NULL.

Дополнительно прочитать о результатах действия трехзначных операторов можно здесь: https://docs.microsoft.com/ru-ru/sql/t-sql/language-elements/null-and-unknown-transact-sql?view=sql-server-ver15

## 2.7. Упорядочение выходных полей (ORDER BY)

Как уже отмечалось, записи в таблицах реляционной базы данных неупорядочены. Однако, данные, выводимые в результате выполнения запроса, могут быть упорядочены. Для этого в операторе SELECT используется предложение с ключевым словом ORDER BY, которое позволяет упорядочивать выводимые записи в соответствии со значениями одного или нескольких выбранных столбцов. При этом можно задать возрастающую (ASC) или убывающую (DESC) последовательность сортировки для каждого из столбцов. По умолчанию принята возрастающая последовательность сортировки. Запрос, позволяющий выбрать все данные из таблицы предметов обучения SUBJECT, с упорядочением по наименованиям предметов, выглядит следующим образом:
```
SELECT *
    FROM SUBJECT
    ORDER BY SUBJ_NAME;
```
Тот же список, но упорядоченный в обратном порядке, можно
получить запросом:
```
SELECT *
    FROM SUBJECT
    ORDER BY SUBJ_NAME DESC;
```
Можно упорядочить выводимый список предметов обучения по значениям семестров, а внутри семестров — по наименованиям предметов:
```
SELECT *
    FROM SUBJECT
    ORDER BY SEMESTR, SUBJ_NAME;
```
Предложение ORDER BY может использоваться с GROUP BY. При этом оператор ORDER BY в запросе всегда должен быть последним:
```
SELECT SUBJ_NAME, SEMESTR, MAX(HOUR)
    FROM SUBJECT
    GROUP BY SEMESTR, SUBJ_NAME
    ORDER BY SEMESTR;
```
При упорядочивании вместо наименований столбцов можно указывать их номера, имея, однако, в виду, что в данном случае это номера
столбцов, указанные при определении выходных данных в запросе, а не номера столбцов в таблице. В запросе
```
SELECT SUBJ_ID, SEMESTR
    FROM SUBJECT
    ORDER BY 2 DESC;
```
выводимые записи будут упорядочены по полю SEMESTR.

Если в поле, которое используется для упорядочивания, существуют NULL-значения, то все они размещаются в конце или предшествуют
всем остальным значениям этого поля.

<img src=./Img/icon-4017417__340.webp style="width:190px;height:160px">
<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания c помощью DB Browser for SQLite</p>

1. Предположим, что стипендия всем студентам увеличена на 20%. Напишите запрос к таблице STUDENT, выполняющий вывод номера студента, его фамилии и величины увеличенной стипендии. Выходные данные упорядочите: а) по значению последнего столбца (величине стипендии); б) в алфавитном порядке фамилий студентов.
2. Напишите запрос, который по таблице EXAM_MARKS позволяет найти а) максимальные и б) минимальные оценки каждого студента и выводит их вместе с идентификатором студента.
3. Напишите запрос, выполняющий вывод списка предметов обучения в порядке а) убывания семестров и б) возрастания количества отводимых на предмет часов. Поле семестра в выходных данных должно быть первым, за ним должны следовать имя предмета обучения и идентификатор предмета.
4. Напишите запрос, который для каждой даты сдачи экзаменов выполняет вывод суммы баллов всех студентов и представляет результаты в порядке убывания этих сумм.
5. Напишите запрос, который для каждой даты сдачи экзаменов выполняет вывод а) среднего, б) минимального, в) максимального баллов всех студентов и представляет результаты в порядке убывания этих значений.

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните указанные задания c помощью Python в интерактивной среде Jupyter notebook</p>

## 2.8. Вложенные подзапросы
SQL позволяет использовать одни запросы внутри других запросов,
т. е. вкладывать запросы друг в друга. Предположим, известна фами-
лия студента (Петров), но неизвестно значение поля STUDENT_ID для
него. Чтобы извлечь данные обо всех оценках этого студента, можно
записать следующий запрос:
```
SELECT *
FROM EXAM_MARKS
WHERE STUDENT_ID =
(SELECT STUDENT_ID
FROM STUDENT SURNAME = "Петров");
```
Следует обратить внимание, что этот запрос корректен только в том случае, если в результате выполнения указанного в скобках подзапроса возвращается единственное значение. Если в результате выполнения подзапроса будет возвращено несколько значений, то при выполнении запроса будет зафиксирована ошибка. В данном примере это произойдет, если в таблице STUDENT будет несколько записей со значениями поля SURNAME = ‘Петров’.

В некоторых случаях для гарантии получения единственного значения в результате выполнения подзапроса используется DISTINCT. Одним из видов функций, которые автоматически всегда выдают в результате единственное значение для любого количества строк, являются агрегирующие функции.

Оператор IN также широко применяется в подзапросах. Он задает список значений, с которыми сравниваются другие значения для определения истинности задаваемого этим оператором предиката. Данные обо всех оценках (таблица EXAM_MARKS) студентов из Воронежа можно выбрать с помощью следующего запроса:
```
SELECT *
    FROM EXAM_MARKS
    WHERE STUDENT_ID IN
        (SELECT STUDENT_ID
        FROM STUDENT
        WHERE CITY = ‘Воронеж’);
```
Подзапросы можно применять внутри предложения HAVING. Пусть требуется определить количество предметов обучения с оценкой, превышающей среднее значение оценки студента с идентификатором 301:
```
SELECT COUNT(DISTINCT SUBJ_ID), MARK
    FROM EXAM_MARKS
    GROUP BY MARK
    HAVING MARK >
    (SELECT AVG(MARK)
        FROM EXAM_MARKS
        WHERE STUDENT_ID = 301);
```

<img src=./Img/icon-4017417__340.webp style="width:180px;height:160px">
<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните задания c помощью DB Browser for SQLite</p>

1. Напишите запрос, выводящий список студентов, получающих максимальную стипендию, отсортировав его в алфавитном порядке по фамилиям.
2. Напишите запрос, выводящий список студентов, получающих стипендию, превышающую среднее значение стипендии.
3. Напишите запрос, выводящий список студентов, обучающихся в Воронеже, с последующей сортировкой по идентификаторам университетов и курсам.
4. Напишите запрос, выводящий список предметов, на изучение которых отведено максимальное количество часов.
5. Напишите запрос, выполняющий вывод имен и фамилий студентов, место проживания которых не совпадает с городом, в котором находится их университет.
6. Напишите запрос, выводящий список университетов, расположенных в Москве и имеющих рейтинг меньший, чем у ВГУ.

<p style="background-color:powderblue; font-size:125%; color: rebeccapurple;">Выполните задания</p>
<p>Выполните указанные задания c помощью Python в интерактивной среде Jupyter notebook</p>